waku/v2/discv5/discover.go

```diff
@@ -192,9 +192,10 @@ func (d *DiscoveryV5) SetHost(h host.Host) {
 
 // only works if the discovery v5 hasn't been started yet.
 func (d *DiscoveryV5) Start(ctx context.Context) error {
-	// compare and swap sets the discovery v5 to `started` state
-	// and prevents multiple calls to the start method by being atomic.
-	if !d.started.CompareAndSwap(false, true) {
+	d.Lock()
+	defer d.Unlock()
+
+	if d.cancel != nil {
```

Comment:
It would be better to release the Lock once the cancel has been assigned, otherwise we are holding the lock for too long as we are invoking listen which is a networking call. 
This way any other routine waiting on the lock would not get blocked for too long only until the context is created.